src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -129,88 +170,65 @@ protected virtual void Dispose(bool disposing)
         {
             if (disposing)
             {
-                this.disposed = true;
-
-                // !!DO NOT!! dispose the semaphore, for the following reasons:
-                // 1. SemaphoreSlim.Dispose is not safe for concurrent use while something is waiting. Calling Dispose
-                //    requires tracking of all outstanding requests for enter/release.
-                // 2. SemaphoreSlim only allocates resources that could need to be be disposed if the
-                //    SemaphoreSlim.AvailableWaitHandle property is accessed. The semaphore field here is private and we
-                //    do not access this property.
-                //
-                ////this.semaphore.Dispose();
-            }
-        }
-
-        /// <summary>
-        /// Requests access to the lock.
-        /// </summary>
-        /// <param name="waitTask">A task that represents a request for the semaphore.</param>
-        /// <returns>A task whose result is a releaser that should be disposed to release the lock.</returns>
-        private Task<Releaser> LockWaitingHelper(Task waitTask)
-        {
-            Requires.NotNull(waitTask, nameof(waitTask));
+                List<WaiterInfo>? waitersCopy = null;
+                lock (this.syncObject)
+                {
+                    this.disposed = true;
 
-            return waitTask.Status == TaskStatus.RanToCompletion
-                ? this.uncontestedReleaser // uncontested lock
-                : waitTask.ContinueWith(
-                    (waiter, state) =>
+                    if (this.waiters.Count > 0)
                     {
-                        // Re-throw any cancellation or fault exceptions.
-                        waiter.GetAwaiter().GetResult();
-
-                        var semaphore = (AsyncSemaphore)state!;
-
-                        if (semaphore.disposed)
+                        waitersCopy = new List<WaiterInfo>(this.waiters.Count);
+                        var ex = new ObjectDisposedException(this.GetType().FullName);
+                        while (this.waiters.First is { } head)
                         {
-                            throw new ObjectDisposedException(semaphore.GetType().FullName);
+                            head.Value.Trigger.TrySetException(ex);
```

Comment:
How important is diagnostics around this to you?  An exception only contains one stack, so by completing each of these tasks with the exact same exception instance, each consumer may see someone else's stack, or intermingled portions of different consumers stacks.  It's safe, though, in that nothing vital will be corrupted, so this is just a decision about how important it is to have coherent stack traces for these consumers... if it's important, you'll want to create a new ODE for each.

Reply:
This exception isn't thrown here. So I expect it has no stack at all. When the Task we're faulting is awaited and the exception is rethrown, is it this very instance that is rethrown? I hadn't thought of that. I'll create a new one for each.